src/compiler/checker.ts

```diff
@@ -2663,7 +2746,170 @@ namespace ts {
+                return unknownType;
+            }
+
+            // If the last arg isn't a function expr, just resolve
+            // its type and define that as the shape of the module
+            let lastArg = <FunctionExpression>callExpr.arguments[callExpr.arguments.length - 1];
+            if (!isFunctionLike(lastArg)) {
+                let resultType = getTypeOfExpression(lastArg);
+                // If this type is a function type, we want to use its return type since
+                // it's going to get invoked anyway
+                let signatures = getSignaturesOfType(resultType, SignatureKind.Call);
+                if (signatures.length > 0) {
+                    return getReturnTypeOfSignature(signatures[0]);
+                }
+                else {
+                    return resultType;
+                }
+            }
+
+            // If there are any expressionful return statements in this function,
+            // we'll use those as the shape of the module
+            let bodyReturnType = getReturnTypeFromBody(lastArg);
+            if (bodyReturnType !== voidType) {
+                return bodyReturnType;
+            }
+
+            // Collect assignments in the body to 'module.exports' or 'exports.propName' or 'module.exports.propName'
+            let assignedModuleType: Type = undefined;
+            let exportAssignedProperties: Symbol[] = [];
+            let exportAssignedPropTable: SymbolTable = {};
+
+            if (cjsModuleType === undefined || cjsExportsType === undefined) {
+                return unknownType;
+            }
+
+            traverse(lastArg.body);
+
+            // Someone assigned to 'module.exports', so ignore other assignments
+            if (assignedModuleType) {
+                return assignedModuleType;
+            }
+
+            // Otherwise return the collected property assignments
+            if (exportAssignedProperties.length > 0) {
+                return createAnonymousType(lastArg.symbol,
+                        exportAssignedPropTable,
+                        /*callSignatures*/ [],
+                        /*constructSignatures*/ [],
+                        /*stringIndexType*/ undefined,
+                        /*numberIndexType*/ undefined);
+            }
+
+            // Empty module?
+            return emptyObjectType;
+
+            function traverse(node: Node) {
+                if (node.kind === SyntaxKind.BinaryExpression &&
+                   (<BinaryExpression>node).operatorToken.kind === SyntaxKind.EqualsToken) {
+                    let lhs = (<BinaryExpression>node).left;
+                    if (lhs.kind === SyntaxKind.PropertyAccessExpression) {
+                        // e.g. module.exports = foo;
+                        let propertyName = (<PropertyAccessExpression>lhs).name.text;
+                        let operandType = getTypeOfNode((<PropertyAccessExpression>lhs).expression);

```

Comment:
Use `checkExpression`. The `getTypeOfNode` function is a helper for the language service.